pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,744 @@
+	}
+
+	return nil, nil
+}
+
+type qemuOCILoader struct{}
+
+func (b qemuOCILoader) LoadImage(ctx context.Context, layer v1.Layer, arch apko_types.Architecture, bc *apko_build.Context) (ref string, err error) {
+	_, span := otel.Tracer("melange").Start(ctx, "qemu.LoadImage")
+	defer span.End()
+
+	// qemu does not have the idea of container images or layers or such, just
+	// create an initramfs from the layer
+	guestInitramfs, err := os.CreateTemp("", "melange-guest-*.initramfs.cpio")
+	if err != nil {
+		return ref, fmt.Errorf("failed to create guest dir: %w", err)
+	}
+
+	// in case of some kernel images, we also need the /lib/modules directory to load
+	// necessary drivers, like 9p, virtio_net which are foundamental for the VM working.
+	if qemuModule, ok := os.LookupEnv("QEMU_KERNEL_MODULES"); ok {
+		clog.FromContext(ctx).Info("qemu: QEMU_KERNEL_MODULES env set, injecting modules in initramfs")
+		if _, err := os.Stat(qemuModule); err == nil {
+			clog.FromContext(ctx).Infof("qemu: local QEMU_KERNEL_MODULES dir detected, injecting")
+			layer, err = injectKernelModules(ctx, layer, qemuModule)
+			if err != nil {
+				clog.FromContext(ctx).Errorf("qemu: could not inject needed kernel modules into initramfs: %v", err)
+				return "", err
+			}
+		}
+	}
+
+	err = apko_cpio.FromLayer(layer, guestInitramfs)
+	if err != nil {
+		return ref, fmt.Errorf("failed to create cpio initramfs: %w", err)
+	}
+
+	return guestInitramfs.Name(), nil
+}
+
+func (b qemuOCILoader) RemoveImage(ctx context.Context, ref string) error {
+	clog.FromContext(ctx).Infof("removing image path %s", ref)
+	return os.RemoveAll(ref)
+}
+
+func createMicroVM(ctx context.Context, cfg *Config) error {
+	clog.FromContext(ctx).Debug("qemu: ssh - create ssh key pair")
+	pubKey, err := generateSSHKeys(ctx, cfg)
+	if err != nil {
+		return err
+	}
+
+	// always be sure to create the VM rootfs first!
+	kernelPath, rootfsInitrdPath, err := getKernelPath(ctx, cfg)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("could not prepare rootfs: %v", err)
+		return err
+	}
+
+	baseargs := []string{}
+	// load microvm profile and bios, shave some milliseconds from boot
+	// using this will make a complete boot->initrd (with working network) In ~700ms
+	// instead of ~900ms.
+	if _, err := os.Stat("/usr/share/qemu/bios-microvm.bin"); err == nil {
+		// only enable pcie for network, enable RTC for kernel, disable i8254PIT, i8259PIC and serial port
+		baseargs = append(baseargs, "-machine", "microvm,rtc=on,pcie=on,pit=off,pic=off,isa-serial=off")
+		baseargs = append(baseargs, "-bios", "/usr/share/qemu/bios-microvm.bin")
+	}
+
+	baseargs = append(baseargs, "-kernel", kernelPath)
+	baseargs = append(baseargs, "-initrd", rootfsInitrdPath)
+
+	if cfg.Memory != "" {
+		memKb, err := convertMemoryToKB(cfg.Memory)
+		if err != nil {
+			return err
+		}
+
+		baseargs = append(baseargs, "-m", fmt.Sprintf("%dk", memKb))
+	} else {
+		baseargs = append(baseargs, "-m", getAvailableMemoryKB())
```

Comment:
this is kind of weird.
`"-m", getAvailableMemoryKB()` reads like "pass all the memory in the system".
versus:

`"-m", getAvailableMemoryKB()/2` or something.

Reply:
Resolved

Reply:
We are intentionally passing '-m' a value that indicates all the system memory.  Am I reading that right?

Even in a single hardware machine that is provisioned to do nothing other than that, we'd be over-committing the memory, right?

Reply:
if nothing is specified it uses available memory
thi shouldn't be a problem at least on linux as qemu only uses whatever it needs, it does not block-allocate all the ram

Else we do 1/2 of available memory? wdyt?

Reply:
I guess we can try whatever and see how it works.

It becomes weird when the only way you can *not* use all the system memory is for the melange yaml file you're using to specify; Generally speaking, that configuration in the melange yaml is saying "I need more resources" but here its the only way to turn things down.

I guess really we should do something like:

```go
requested := cfg.Memory
system := getAvailableMemoryKB()
if requested > 0 && requested > system  {
   panic("BAD")
}
baseargs = append(baseargs, "-m", system)
```

Reply:
Perfect, defaulted to 1/4th of available memory, and check if cfg.mem > available mem and error out :+1: